user-data-store-server/src/main/java/com/wultra/security/userdatastore/service/ClaimsService.java

```diff
@@ -0,0 +1,200 @@
+                    logger.debug("Updating claims of user ID: {}, stored claim: {}", userId, claim);
+                    final String claims = encryptionService.decryptDocumentData(entity);
+                    final Map<String, Object> claimMap;
+                    try {
+                        claimMap = objectMapper.readValue(claims, new TypeReference<>() {});
+                        claimMap.put(claim, value);
+                        encryptionService.encryptDocumentData(entity, objectMapper.writeValueAsString(claimMap));
+                    } catch (JsonProcessingException e) {
+                        throw new InvalidRequestException(e);
+                    }
+                    entity.setTimestampLastUpdated(LocalDateTime.now());
+
+                    documentRepository.save(entity);
+                    audit("Updated claims of user ID: {}, stored claim: {}", userId, claim);
+                },
+                () -> {
+                    logger.debug("Creating new claims of user ID: {}, stored claim: {}", userId, claim);
+                    final DocumentEntity entity = new DocumentEntity();
+                    entity.setId(UUID.randomUUID().toString());
+                    entity.setUserId(userId);
+                    entity.setDocumentType(CLAIMS_DOCUMENT_TYPE);
+                    entity.setDataType(CLAIMS_DATA_TYPE);
+                    entity.setDocumentDataId(CLAIMS_DOCUMENT_DATA_ID);
+                    entity.setAttributes("{}");
+                    entity.setTimestampCreated(LocalDateTime.now());
+
+                    final Map<String, Object> claimMap = new LinkedHashMap<>();
+                    claimMap.put(claim, value);
+                    try {
+                        encryptionService.encryptDocumentData(entity, objectMapper.writeValueAsString(claimMap));
+                    } catch (JsonProcessingException e) {
+                        throw new InvalidRequestException(e);
+                    }
+
+                    documentRepository.save(entity);
+                    audit("Created claims for user ID: {}, stored claim: ", userId, claim);
+                });
+    }
+
+    @Transactional
+    public void createOrUpdateClaims(final String userId, final Object claims) {
+        final String claimsAsString;
+        try {
+            claimsAsString = objectMapper.writeValueAsString(claims);
+        } catch (JsonProcessingException e) {
+            throw new InvalidRequestException(e);
+        }
+        documentRepository.findAllByUserIdAndDataType(userId, CLAIMS_DATA_TYPE).stream().findAny()
+                .ifPresentOrElse(entity -> {
+                            logger.debug("Updating claims of user ID: {}", userId);
+                            encryptionService.encryptDocumentData(entity, claimsAsString);
+                            entity.setTimestampLastUpdated(LocalDateTime.now());
+                            audit("Updated claims of user ID: {}", userId, null);
```

Comment:
Yes, this is quite hard to audit, because we receive an `Object` and we do not want to audit the claim values, they can contain sensitive records. I will revise the logic for auditing though.